### ISSUE TITLE
`modernize-use-emplace` not reported when accessing container through pointer

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#55869]
```cpp
#include <list>
#include <string>

static void f(std::list<std::string>* l)
{
    l->push_back(""); // no warning
}
```

https://godbolt.org/z/3bcT6ccxx

The following invalid code will report the warning though:
```cpp
#include <list>
#include <string>

static void f(std::list<std::string> l)
{
    l->push_back(""); // warning
}
```
https://godbolt.org/z/s14GfY7G5

**Comment by JonasToth** [comment]
Could be addressed together with #55856 and #55870

**Comment by nicovank** [comment]
I actually ran into this same issue working on another check (with the same principle of matching on certain functions in certain containers). This is how I fixed it then.

```cpp
const auto hasSoughtArrowMemberExpr = has(memberExpr(
    isArrow(),
    hasObjectExpression(
        hasType(pointerType(pointee(hasUnqualifiedDesugaredType(
            recordType(hasDeclaration(soughtContainerDecl)))))))));

// We don't need to make sure the MemberExpr is not an arrow call, if it was
// the code wouldn't compile anyway.
const auto hasSoughtPeriodMemberExpr =
    has(memberExpr(hasObjectExpression(hasType(hasUnqualifiedDesugaredType(
        recordType(hasDeclaration(soughtContainerDecl)))))));

Finder->addMatcher(
    cxxMemberCallExpr(
        anyOf(hasSoughtArrowMemberExpr, hasSoughtPeriodMemberExpr),
        ...
```

**Comment by BigPeet** [comment]
I've analyzed this a bit: The issue seems to be that all the cxxMemberCallExprs are filtered based on "on(hasType(...)".
E.g.
```cpp
auto CallPushBack = cxxMemberCallExpr(
      hasDeclaration(functionDecl(hasName("push_back"))),
      on(hasType(hasCanonicalType(
          hasDeclaration(cxxRecordDecl(hasAnyName(ContainersWithPushBack)))))));
```

This does not properly catch "pointerType". We would need something like this:

```cpp
auto CallPushBack = cxxMemberCallExpr(
      hasDeclaration(functionDecl(hasName("push_back"))),
      anyOf(
          on(hasType(hasCanonicalType(hasDeclaration(cxxRecordDecl(hasAnyName(ContainersWithPushBack))))))),
          on(hasType(pointerType(pointee(hasCanonicalType(hasDeclaration(cxxRecordDecl(hasAnyName(ContainersWithPushBack))))))))
     ),
);
```

I will take a swing at this.

**Comment by nicovank** [comment]
That looks about right. If you submit a patch you can add me as reviewer. I don't have commit access but will review and stamp.

**Comment by BigPeet** [comment]
Submitted a patch: https://reviews.llvm.org/D142939
